File: Day1_PointerBasics_Go/hard/hard_04.go
// FILE: hard_04.go
// TOPIC: Pointer Basics - Day 1 (Go)
// DIFFICULTY: Hard
//
// TASK: Go allows returning a pointer to a local variable (heap escape)
// ---------------------------------------------------------------------
// In C++, returning the address of a local variable is undefined behaviour
// because the local is destroyed when the function returns.
// In Go, the COMPILER detects this and automatically allocates the variable
// on the HEAP, keeping it alive as long as the pointer exists.
// This is called "escape to heap".
//
// 1. Write a function:
//      func newInt(val int) *int
//    Inside, declare  n := val  and return &n.
//    (In C++ this would be UB. In Go it is perfectly safe.)
//
// 2. In main():
//      a) Call newInt(10) and store the returned pointer in `p`.
//      b) Print *p.
//      c) Modify the value through p (*p = 99).
//      d) Print *p again.
//      e) Call newInt(55) and store in a second pointer `q`.
//      f) Print *p and *q — they are INDEPENDENT heap allocations.
//
// EXPECTED OUTPUT:
//   *p after alloc  : 10
//   *p after modify : 99
//   *p=99, *q=55  (independent)

package main

import "fmt"

// TODO: Write newInt here.

func main() {
	// TODO: Write your solution here.

}

File: Day1_PointerBasics/hard/hard_04.cpp
/*
 * FILE: hard_04.cpp
 * TOPIC: Pointer Basics - Day 1
 * DIFFICULTY: Hard
 *
 * TASK: const pointer vs pointer to const
 * ----------------------------------------
 * There are THREE distinct pointer+const combinations. Understand each:
 *
 *   (A) int* const p        -- constant pointer to int
 *                              p cannot be re-assigned; *p CAN be changed.
 *   (B) const int* p        -- pointer to constant int
 *                              p CAN be re-assigned; *p cannot be changed.
 *   (C) const int* const p  -- constant pointer to constant int
 *                              neither p nor *p can change.
 *
 * 1. Declare:  int a = 1, b = 2;
 *
 * 2. Case A: int* const pA = &a;
 *      - Try to change *pA (change a's value) - this should COMPILE fine.
 *      - Do NOT attempt to re-assign pA itself (that would be a compile error).
 *      - Print *pA after the change.
 *
 * 3. Case B: const int* pB = &a;
 *      - Re-assign pB to point to b - this should COMPILE fine.
 *      - Do NOT attempt to modify *pB (that would be a compile error).
 *      - Print *pB after re-assignment.
 *
 * 4. Case C: const int* const pC = &b;
 *      - Just print *pC.
 *      - Do NOT attempt to change either pC or *pC.
 *
 * EXPECTED OUTPUT:
 *   Case A (*pA after write)  : 10
 *   Case B (*pB after re-point): 2
 *   Case C (*pC)              : 2
 */

#include <iostream>
using namespace std;

int main() {
    // TODO: Write your solution here.

    int a = 1, b = 2;

    int* const pA = &a;
    *pA = 10;
    cout << "(*pA after write) :" << *pA << "\n";

    const int* pB = &a;
    pB = &b;
    cout << "(*pB after re-point): " << *pB << "\n";

    const int* const pC = &b;
    cout << "(*pC) :" << *pC << "\n";


    return 0;
}
